packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts

```diff
@@ -0,0 +1,192 @@
+import type {
+  DocRecord,
+  DocsService,
+  WorkspaceLocalState,
+  WorkspaceService,
+} from '@toeverything/infra';
+import { Entity, LiveData } from '@toeverything/infra';
+import type { Observable } from 'rxjs';
+import { race, skip, Subject } from 'rxjs';
+
+import { resolveLinkToDoc } from '../../navigation';
+
+type QuickSearchMode = 'cmdk' | 'search-callback';
+
+const RECENT_PAGES_LIMIT = 3; // adjust this?
+const RECENT_PAGES_KEY = 'recent-pages';
+
+const EMPTY_ARRAY: string[] = [];
+
+export type SearchCallbackResult =
+  | {
+      docId: string;
+      blockId?: string;
+    }
+  | {
+      query: string;
+      action: 'insert';
+    };
+
+// todo: move command registry to entity as well
+export class QuickSearchManager extends Entity {
+  constructor(
+    private readonly localState: WorkspaceLocalState,
+    private readonly docsService: DocsService,
+    private readonly workspaceService: WorkspaceService
+  ) {
+    super();
+  }
+  private showCounter = 0;
+  private readonly _mode$ = new LiveData<{
+    counter: number;
```

Comment:
It seems unnecessary to record this.

If only one quicksearch is displayed at a time, we can hide the older one before showing new one.